python/spark_sklearn/keyed_models.py

```diff
@@ -0,0 +1,407 @@
+"""
+Classes for implementing keyed models.
+
+The use case that this addresses is where a client has a dataset with many keys - the distribution
+ of distinct keys amongst the is such that the total number of rows for every key value can be
+contained completely in memory on a single machine.
+
+This assumption is particularly enabling because clients may wish to apply more intricate single-
+machine models (such as a scikit-learn estimator) to every user.
+
+The API provided here generalizes the scikit-learn estimator interface to the Spark ML one; in
+particular, it allows clients to train their scikit-learn estimators in parallel over a grouped
+and aggregated dataframe.
+
+>>> import numpy as np
+>>> from sklearn.linear_model import LinearRegression
+>>> from pyspark.ml.linalg import Vectors
+>>> from pyspark.sql import SparkSession
+>>> from pyspark.sql.functions import udf
+>>> spark = SparkSession.builder.master("local").getOrCreate()
+>>> df = spark.createDataFrame([(user,
+...                              Vectors.dense([i, i ** 2, i ** 3]),
+...                              user + i + 2 * i ** 2 + 3 * i ** 3)
+...                             for user in range(3) for i in range(5)])
+>>> df = df.toDF("key", "features", "y")
+>>> km = KeyedEstimator(sklearnEstimator=LinearRegression(), yCol="y").fit(df)
+>>> printnp = lambda nparr: "[" + ", ".join("{:.2f}".format(x) for x in nparr) + "]"
```

Comment:
this works, but this style is rather foreign in python, I find. Can you used the `def` syntax?